### PR TITLE
Support for PHP 7.2's `object` hint

### PR DESCRIPTION
**Pull request by Ocramius**
This patch ensures that the `object` type hint introduced by PHP 7.2 is correctly analysed and treated as a special type.

Also, I imported the current state of php-src reflection tests

### CODE REVIEW

File: src/Reflection/ReflectionParameter.php [hunk 1]
```diff
@@ -541,11 +541,17 @@ public function getClass() : ?ReflectionClass
             return null;
         }
 
+        $fqsen = $hint->getFqsen();
+
+        if (! $fqsen) {
```
Comment (Ocramius) [inline src/Reflection/ReflectionParameter.php]: the cyclomatic complexity:

too-damn-high

File: test/core/002.phpt [hunk 2]
```diff
@@ -50,14 +50,14 @@ var_dump($r->bar);
 ===DONE===
 --EXPECTF--
 ReflectionMethodEx::__construct
-%unicode|string%(26) "ReflectionFunctionAbstract"
-%unicode|string%(7) "getName"
-%unicode|string%(3) "xyz"
+string(26) "ReflectionFunctionAbstract"
```
Comment (asgrim) [inline test/core/002.phpt]: Why changing core `phpt` tests here? They're not strictly relevant, can this be done in a separate PR please if they need updating (which they probably do, to be fair)...

File: test/unit/Reflection/ReflectionMethodTest.php [hunk 3]
```diff
@@ -203,6 +204,25 @@ public function someMethod() {}
...
+    {
+        $php = '<?php
+        namespace A;
+        class Foo {
+            public function someMethod() : object {}
+        }
+        ';
+
+        $returnType = (new ClassReflector(new StringSourceLocator($php)))
+            ->reflect('A\Foo')
```
Comment (asgrim) [inline test/unit/Reflection/ReflectionMethodTest.php]: `::class` can be used

File: test/unit/Reflection/ReflectionParameterTest.php [hunk 4]
```diff
@@ -595,4 +595,23 @@ public function testGetClassFromParentTypeHintedProperty() : void
         self::assertInstanceOf(ReflectionClass::class, $hintedClassReflection);
         self::assertSame('stdClass', $hintedClassReflection->getName());
     }
+
+    public function testGetClassFromObjectTypeHintedProperty() : void
+    {
+        $content = '<?php class Foo { public function myMethod(object $param) {} }';
+
+        $parameter = (new ClassReflector(new StringSourceLocator($content)))
+            ->reflect('Foo')
```
Comment (asgrim) [inline test/unit/Reflection/ReflectionParameterTest.php]: `::class`

File: test/core/002.phpt (hunk 2 above)
Comment (Ocramius) [inline test/core/002.phpt]: I was actually looking for 7.2 tests added around `object`, and figured that replacing them as per current state would be much easier

**Comment by Ocramius** [comment]
Merging as per @asgrim's review.
